rewrite-gradle/src/main/java/org/openrewrite/gradle/internal/AddDependencyVisitor.java

```diff
@@ -115,6 +115,19 @@ public class AddDependencyVisitor extends JavaIsoVisitor<ExecutionContext> {
     public J.Block visitBlock(J.Block block, ExecutionContext ctx) {
         J.Block b = super.visitBlock(block, ctx);
         if (b == block && insertPredicate != null && insertPredicate.test(getCursor())) {
+            // If we're visiting a gradle kotlin file but we're not visiting the highest level block we should not add a dependencies block.
+            if (getCursor().firstEnclosing(K.CompilationUnit.class) != null &&
+                !(getCursor().getParentOrThrow().getValue() instanceof K.CompilationUnit)) {
+                try {
+                    getCursor().dropParentUntil(value -> value instanceof J.MethodInvocation &&
+                                                         ("testing".equals(((J.MethodInvocation) value).getSimpleName()) ||
+                                                         "sourceSets ".equals(((J.MethodInvocation) value).getSimpleName()) ||
+                                                         "configurations ".equals(((J.MethodInvocation) value).getSimpleName()))
+                            );
+                } catch (IllegalStateException e) {
+                    return b;
+                }
+            }
```

Comment:
So I wonder if it may be better to fine tune the `isTopLevel` check within `gradle.AddDependency` rather than including those additional rules here.

You'll have a bit more control there as well since that recipe is _only_ designed to add a dependency to the immediate project's `dependencies` block or add a block if one is missing.